api/v1beta1/keystoneendpoint_types.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// KeystoneEndpointSpec defines the desired state of KeystoneEndpoint
+type KeystoneEndpointSpec struct {
+	// +kubebuilder:validation:Required
+	// ServiceName - Name of the service to create the endpoint for
+	ServiceName string `json:"serviceName,omitempty"`
+	// +kubebuilder:validation:Required
+	// Endpoints - map with service api endpoint URLs with the endpoint type as index
+	Endpoints map[string]string `json:"endpoints,omitempty"`
+}
+
+// KeystoneEndpointStatus defines the observed state of KeystoneEndpoint
+type KeystoneEndpointStatus struct {
+	EndpointIDs map[string]string `json:"endpointIDs,omitempty"`
+	ServiceID   string            `json:"serviceID,omitempty"`
+	// Conditions
+	Conditions condition.Conditions `json:"conditions,omitempty" optional:"true"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.conditions[0].status",description="Status"
+//+kubebuilder:printcolumn:name="Message",type="string",JSONPath=".status.conditions[0].message",description="Message"
+
+// KeystoneEndpoint is the Schema for the keystoneendpoints API
+type KeystoneEndpoint struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KeystoneEndpointSpec   `json:"spec,omitempty"`
+	Status KeystoneEndpointStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// KeystoneEndpointList contains a list of KeystoneEndpoint
+type KeystoneEndpointList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KeystoneEndpoint `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&KeystoneEndpoint{}, &KeystoneEndpointList{})
+}
+
+// IsReady - returns true if endpoint got created ok in keystone
+// AND the endpoint ID registerd in the object status
+func (instance KeystoneEndpoint) IsReady() bool {
+
+	return instance.Status.Conditions.IsTrue(KeystoneServiceOSEndpointsReadyCondition) &&
+		len(instance.Status.EndpointIDs) == len(instance.Spec.Endpoints)
```

Comment:
I'm not sure we need other than the check for the ReadyCondition. But if we need to check on the Endpoints as there might be non reconciled changes between Spec and Status there, then a simple length equality is not enough. What if the URL changed within an endpoint in the Spec? The len() check does not detect that.

Reply:
I think right now that `status.EndpointIDs` stores a map of endpoint "type" (public, internal, admin) to an ID in Keystone, while the `spec.Endpoints` is a map of type to URL.  So if we wanted to compare those URLs to something in the status, perhaps we'll need to store the URLs in the status as well.  Basically right now can only compare endpoint types presence and absence.

Reply:
correct, as Andrew mentioned, right now its a map type -> endpoint ID. when an endpoint gets removed from the spec, the controller deletes the endpoint from keystone and when successful, also the ID gets deleted from the status [1]. So if we have the case with 2 endpoints, when one gets replaced and the delete of the endpoint is failing (for whatever reason), the spec number is 2 and the status.EndpointIDs also has count of 2. But the KeystoneServiceOSEndpointsReadyCondition will catch this. When an error happens during that delete, the KeystoneServiceOSEndpointsReadyCondition flips to false.

[1] https://github.com/openstack-k8s-operators/keystone-operator/pull/100/files#diff-33a1e743ae4c5cb0d0dad9e55e003bc15061dc9c169ad554a5368a956859102aR349

Reply:
If reconciliation error flips the KeystoneServiceOSEndpointsReadyCondition then that also support my point that the check for length is not needed here. 

It all depends on how precise we want to be with the IsReady().

- a) We want to say IsReady() -> False if the Spec.Endpoints was updated but the reconcile loop hasn't been triggered yet. Then we need to compare Spec.Endpoints with Status.Endpoints here to be able to detect if there was a change that needs reconciliation. But if we want to do this then simply checking the length is not enough as might be the URL changed in an existing Endpoint. I understand that this is not possible today as the Status.Endpoint does not have the URL.

- b) We are OK to keep IsReady() -> True if the Spec.Endpoints was updated but the reconcile loop hasn't been triggered yet. In this case the 
```
IsReady() {return instance.Status.Conditions.IsTrue(KeystoneServiceOSEndpointsReadyCondition)}
```
is enough.

Reply:
32e45f2bfe61883678747b03c2257f67f93f15da